image-customize

```diff
@@ -82,6 +83,114 @@ class InstallAction(ActionBase):
 
 
+class BuildAction(ActionBase):
+    '''Build and install distribution package(s) from dist tarball'''
+
+    quick = False
+    verbose = False
+
+    @staticmethod
+    def execute(machine_instance, tarball):
+        # upload the tarball
+        tarname = os.path.basename(tarball)
+        vm_tar = os.path.join("/var/tmp", tarname)
+        # looks like cockpit-foo-123.10.gdbde381.tar.xz
+        m = re.search(r'([^-]+)\.tar(\..*)?$', tarname)
+        if not m:
+            raise ValueError(f"cannot parse version number from tarball {tarname}")
+        version = m.group(1)
+        machine_instance.upload([tarball], vm_tar, relative_dir=".")
+
+        machine_instance.execute("rm -rf /var/tmp/build; mkdir -p /var/tmp/build")
+
+        # this will fail if neither is available -- exception is clear enough, this is a developer tool
+        out = machine_instance.execute("(which pbuilder || which mock || which pacman) 2>/dev/null")
+        if 'pbuilder' in out:
+            BuildAction.build_deb(machine_instance, vm_tar, version)
+        elif 'mock' in out:
+            BuildAction.build_rpm(machine_instance, vm_tar, version)
+        elif 'pacman' in out:
+            BuildAction.build_arch(machine_instance, vm_tar, version)
+        else:
+            raise NotImplementedError(f"unknown build platform: {out}")
+
+        # machine_instance.execute(f"{install_command} {package}", timeout=1800)
+
+    def build_deb(machine, vm_tar, version):
+        build_opts = 'nocheck' if BuildAction.quick else ''
+
+        # build source packge
+        machine.execute(f"""
+            set -eu
+            export DEB_BUILD_OPTIONS='{build_opts}'
+            tar -C /var/tmp/build -xf '{vm_tar}'
+            cd "$(ls -d /var/tmp/build/*)"
+            # find and copy debian packaging directory
+            cp -r "$(dirname $(find -path '*/debian/control'))" .
+            # create orig.tar link for building dsc
+            source=$(awk '/^Source: / {{ print $2 }}' debian/control)
+            ln -s '{vm_tar}' ../${{source}}_{version}.orig.tar.xz
+            # fill in the version number
+            sed -i '1 s/([^)-]*/({version}/' debian/changelog
+
+            dpkg-buildpackage -S -us -uc -nc""")
+
+        # build binary packages
+        machine.execute("cd /var/tmp/build; pbuilder build --buildresult . *.dsc", timeout=1800)
+
+        # install packages
+        machine.execute("dpkg -i /var/tmp/build/*.deb")
+
+    def build_rpm(machine, vm_tar, version):
+        mock_opts = ''
+        if BuildAction.verbose:
+            mock_opts += ' --verbose'
+        if BuildAction.quick:
+            mock_opts += ' --nocheck'
+
+        # HACK: mock has *_builddep_command config options for all backends *except* dnf
+        # dnf builddep insists on a cache, which we don't need as the mock is offline, pre-installed,
+        # and single-use only
+        machine.execute('''sed -i '/self.builddep_command =/ s_self.command_"/bin/true"_' '''
```

Comment:
I found that `mock --disablerepo=updates --offline` was enough to make it stop trying to update the selinux packages (which I think was the only issue).

Reply:
I was really hoping that this bug got fixed in the recent years, but alas it wasn't. Without it:

```
 # /usr/bin/dnf builddep --installroot /var/lib/mock/fedora-35-x86_64/root/ --releasever 35 -C --setopt=deltarpm=False --allowerasing --disableplugin=local --disableplugin=spacewalk --disableplugin=versionlock /var/lib/mock/fedora-35-x86_64/root//builddir/build/SRPMS/cockpit-podman-40.13.ga3acf22-1.fc35.src.rpm --setopt=tsflags=nocontexts
No matches found for the following disable plugin patterns: local, spacewalk, versionlock
Updating Subscription Management repositories.
Unable to read consumer identity

This system is not registered with an entitlement server. You can use subscription-manager to register.

Error: Cache-only enabled but no cache for 'fedora'
```

Reply:
Can we also `--disablerepo=fedora` then?  Or just not delete the cache...

Reply:
`--disablerepo=*` works, tested with cockpit-podman and cockpit. That's a bit less gross.

(FTR, that hack originated in commit 847b2043 two years ago, it's far from new)